drivers/serial/uart_nrfx_uarte2.c

```diff
@@ -0,0 +1,1058 @@
+			&CONCAT(a2i_config_, idx) : NULL, \
+		.nrfx_dev = NRFX_UARTE_INSTANCE(idx), \
+		.nrfx_config = { \
+			.p_context = (void *)DEVICE_DT_GET(UARTE(idx)), \
+			.tx_cache = { \
+				.p_buffer = UART_CACHE_VAR(TX, idx), \
+				.length = UART_CACHE_SIZE(TX, idx) \
+			}, \
+			.rx_cache = { \
+				.p_buffer = UART_CACHE_VAR(RX, idx), \
+				.length = UART_CACHE_SIZE(RX, idx) \
+			}, \
+			.p_rx_cache_scratch = &CONCAT(uarte, idx, rx_cache_scratch), \
+			.baudrate = NRF_BAUDRATE(CONCAT(uart, idx, _config).baudrate), \
+			.interrupt_priority = DT_IRQ(UARTE(idx), priority), \
+			.config = NRF_UARTE_CONFIG(CONCAT(uart, idx, _config), \
+						UARTE_ODD_PARITY_ALLOWED), \
+			.tx_stop_on_end = \
+				IS_ENABLED(CONCAT(CONFIG_UART_, idx, _ENHANCED_POLL_OUT)), \
+			.skip_psel_cfg = true, \
+			.skip_gpio_cfg = true, \
+		}, \
+		.pcfg = PINCTRL_DT_DEV_CONFIG_GET(UARTE(idx)), \
+		.config = { \
+			.baudrate = UARTE_PROP(idx, current_speed), \
+			.parity = IS_ENABLED(CONCAT(CONFIG_UART_, idx, _NRF_PARITY_BIT)) ? \
+				  UART_CFG_PARITY_EVEN : UART_CFG_PARITY_NONE, \
+			.stop_bits = UART_CFG_STOP_BITS_1, \
+			.data_bits = UART_CFG_DATA_BITS_8, \
+			.flow_ctrl = UARTE_PROP(idx, hw_flow_control) ?  \
+				     UART_CFG_FLOW_CTRL_RTS_CTS : UART_CFG_FLOW_CTRL_NONE, \
+		}, \
+		.flags = (UARTE_HAS_PROP(idx, rx_pin) ?	\
+				0 : UARTE_CFG_FLAG_NO_RX) | \
+			(IS_ENABLED(CONCAT(CONFIG_UART_, idx, _GPIO_MANAGEMENT)) ? \
+				UARTE_CFG_FLAG_GPIO_MGMT : 0) | \
+			(IS_ENABLED(CONCAT(CONFIG_UART_, idx, _INTERRUPT_DRIVEN)) ? \
+				UARTE_CFG_FLAG_INTERRUPT_DRIVEN_API : 0), \
+		LOG_INSTANCE_PTR_INIT(log, LOG_MODULE_NAME, idx) \
+	}; \
+	static struct uart_async_to_irq_data CONCAT(a2i_data_, idx); \
+	static struct uarte_async_data CONCAT(uarte, idx, _async); \
+	static struct uarte_nrfx_data CONCAT(uarte_, idx, _data) = { \
+		.a2i_data = IS_ENABLED(CONCAT(CONFIG_UART_, idx, _INTERRUPT_DRIVEN)) ? \
+			&CONCAT(a2i_data_, idx) : NULL, \
+		.async = (IS_ENABLED(CONCAT(CONFIG_UART_, idx, _INTERRUPT_DRIVEN)) || \
+			  IS_ENABLED(CONCAT(CONFIG_UART_, idx, _ASYNC))) ? \
+				 &CONCAT(uarte, idx, _async) : NULL \
+	}; \
+	static int CONCAT(uarte_, idx, _init)(const struct device *dev) \
+	{ \
+		(void)uarte_nrfx_pm_action;/* To avoid compilation warnings. */ \
+		COND_CODE_1(INSTANCE_POLLING(_, /*empty*/, idx, _), (), \
+			( \
+			IRQ_CONNECT(DT_IRQN(UARTE(idx)), \
+				    DT_IRQ(UARTE(idx), priority), \
+				    nrfx_isr, CONCAT(nrfx_uarte_, idx, _irq_handler), 0); \
+			irq_enable(DT_IRQN(UARTE(idx))); \
+			)\
+		)\
+		return uarte_nrfx_init(dev); \
+	} \
+	PM_DEVICE_DT_DEFINE(UARTE(idx), uarte_nrfx_pm_action); \
+	IF_ENABLED(CONCAT(CONFIG_UART_, idx, _INTERRUPT_DRIVEN), \
+		(static int CONCAT(uart_rx_start, idx)(void) { \
+			static const struct device *dev = DEVICE_DT_GET(UARTE(idx)); \
+			start_rx(dev); \
+			return 0; \
+		} \
+		SYS_INIT(CONCAT(uart_rx_start, idx), POST_KERNEL, 0);)) \
```

Comment:
Why is this done here, not in `uarte_nrfx_init()`?

Reply:
Removed. I think that it got back to times then a2i adaptation layer was using some kernel calls and required that init function is called post kernel but it is not longer the case.